### ISSUE TITLE
[NEED HELP] Evangelise async design

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10682]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)| @puzzlement reported|
|-|-|
|Trac ID|trac#1020|
|Type|enhancement|
|Created|2005-05-08 17:12:54Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1020 1020
type__enhancement enhancement
reporter__hypatia hypatia
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__core core
keywords__documentation documentation
time__1115572374000000 1115572374000000
changetime__1294432540000000 1294432540000000
version__ 
owner__ 
cc__itamarst cc__Predictive cc__kragen cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
This is one of the suggestions from issue 1015.

"Answer the question; 'What's the point of going to all this trouble for an
async design when 'regular' coding is so much easier?' Even a simple link to
the C10k page would help ('Why Use Twisted' is a good spot for it).  Lots of
people hear that Twisted is a great networking toolkit, but come expecting
something more akin to Net::X modules on CPAN and not a plunge into learning
about asynchronous design."

This is actually not an area I have expertise in. There is some scattered
justification in the docs that seems to centre around race conditions, deadlocks
and the GIL. I'll see what I can come up with in unifying this. Suggestions
appreciated.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
1. scalability; see c10k page, plus python maks it worse since threads don't run
in parallel ("GIL") so don't get benefits from multiple CPUs.
2. lack of race conditions.
3. Some protocols are really hard to do with blocking APIs (e.g. think about how
you'd implement chat server if you had only threads and blocking read and write).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Or well, less race conditions. Makes concurrency a much easier problem since you
can;t be interrupted at any point, only at event handler borders (or in defgen
when yielding Deferred).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Thanks Itamar, that should be enough for a respectable write-up.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I hope that wasn't sarcasm? :) I was going to add more detail later if needed.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Not sarcasm! The more info I have the easier it is, but I should have the minium
required now to write something sensible.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="kragen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kragen commented|
|-|-|

```
#!html
<pre>
Is there a URL where the result of this effort will appear?

The best thing I've found so far along these lines is Zooko's 2001 article, "Why
Multithreading Sucks, Using Mojo Nation As An Example," at
http://www.eros-os.org/pipermail/e-lang/2001-July/005410.html
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set status to `closed`|
|-|-|

Not really a problem with Twisted as such, more an issue with programmers' brains.

This should just be done as necessary in the various documents that describe how the reactor works.  I don't see how we could ever close this ticket.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
